### ISSUE TITLE
Организация связей роутеров с внешним миром

### ISSUE THREAD

**Issue by auryadov-funcorp** [theIlyaM/Demo#2]
Считаю, что нужно избавиться от класса BaseRouter. 
Почему: им мы не решаем проблему предыдущего подхода, а по сути делаем то же самое. В чем проблема: это архитектурная система, основанная на наследовании. Есть некая монструозная "правильная" система, правила работы которой закреплены, и она может лишь расширяться в нужных точках. Со временем таких точек становится все больше, размер и сложность системы растут, т.к. она должна все учитывать. Со временем ее начинает не хватать под текущие задачи, и программисты выдумывают костыли, а поддержку системы забрасывают. Рефакторить систему, основанную на наследовании, сложнее всего.
Чтобы ее решить, предлагаю отказаться от использования жесткой структуры (роутеры внутри строятся каждый кастомно), а выносить в общие вещи только то, что реиспользуется. Так лишней сложности не будет, и в система будет жить только то, что реально используется.

Для связи роутеров друг с другом предлагаю использовать прямые зависимости классов роутеров. Например, роутер комментариев должен открывать профиль - он имеет прямую зависимость от роутера профиля (чтобы вызывать его методы). Зависимости конфигурируются в DI.

**Comment by theIlyaM** [comment]
В BaseRouter **конструктор** и два метода: **открыть** и **закрыть** модуль.

Ты хочешь его копировать в каждом роутере?

**Comment by auryadov-funcorp** [comment]
Нет, я предлагаю сделать его отдельным классом и инжектить в каждый роутер. По факту состояние (внутренние переменные) BaseRouter не нужно остальным роутерам. Тогда нет смысла использовать наследование.

**Comment by auryadov-funcorp** [comment]
В результате решил не композировать классы. Почему: интерфейс Router имеет метод close - и это та вещь, которая (думаю) будет использоваться одинаково для всех. Т.е. текущая реализация позволяет закрыть любой роутер без необходимости описывать в каждом классе. Если сделать композицию (как я предлагаю), это улучшит гибкость роутеров (и уменьшит связность), но уберет эту возможность обрабатывать их одинаково. 
Поэтому вывод, что будем не позволять BaseRouter-у расти, чтобы не усложнялся и не протухал, но оставим.
